mmcv/ops/csrc/pytorch/deform_conv_cpu.cpp

```diff
@@ -0,0 +1,724 @@
+                                                                    dilation_h, dilation_w, channel_per_deformable_group, parallel_imgs,
+                                                                    2 * ksize_h * ksize_w * deformable_group, deformable_group,
+                                                                    height_col, width_col, grad_offset_);
+                                                              }));
+}
+
+void DeformConvForwardCPULauncher(Tensor input, Tensor weight,
+                                  Tensor offset, Tensor output,
+                                  Tensor columns, Tensor ones, int kW,
+                                  int kH, int dW, int dH, int padW,
+                                  int padH, int dilationW, int dilationH,
+                                  int group, int deformable_group,
+                                  int im2col_step)
+{
+
+  deform_conv_shape_check(input, offset, NULL, weight, kH, kW, dH, dW, padH,
+                          padW, dilationH, dilationW, group, deformable_group);
+  at::DeviceGuard guard(input.device());
+
+  int batch = 1;
+  if (input.ndimension() == 3)
+  {
+    // Force batch
+    batch = 0;
+    input.unsqueeze_(0);
+    offset.unsqueeze_(0);
+  }
+
+  long batchSize = input.size(0);
+  long nInputPlane = input.size(1);
+  long inputHeight = input.size(2);
+  long inputWidth = input.size(3);
+
+  long nOutputPlane = weight.size(0);
+
+  long outputWidth = (inputWidth + 2 * padW - (dilationW * (kW - 1) + 1)) / dW + 1;
+  long outputHeight = (inputHeight + 2 * padH - (dilationH * (kH - 1) + 1)) / dH + 1;
+
+  TORCH_CHECK((offset.size(0) == batchSize), "invalid batch size of offset");
+
+  output = output.view({batchSize / im2col_step, im2col_step, nOutputPlane, outputHeight, outputWidth});
+  columns = at::zeros({nInputPlane * kW * kH, im2col_step * outputHeight * outputWidth}, input.options());
+
+  if (ones.ndimension() != 2 ||
+      ones.size(0) * ones.size(1) < outputHeight * outputWidth)
+  {
+    ones = at::ones({outputHeight, outputWidth}, input.options());
+  }
+
+  input = input.view({batchSize / im2col_step, im2col_step, nInputPlane,
+                      inputHeight, inputWidth});
+  offset =
+      offset.view({batchSize / im2col_step, im2col_step,
+                   deformable_group * 2 * kH * kW, outputHeight, outputWidth});
+
+  Tensor output_buffer = at::zeros({batchSize / im2col_step, nOutputPlane,
+                                    im2col_step * outputHeight, outputWidth},
+                                   output.options());
+
+  output_buffer = output_buffer.view(
+      {output_buffer.size(0), group, output_buffer.size(1) / group,
+       output_buffer.size(2), output_buffer.size(3)});
+
+  for (int elt = 0; elt < batchSize / im2col_step; elt++)
+  {
+
+    deformable_im2col_cpu(input[elt], offset[elt], nInputPlane, inputHeight,
```

Comment:
If the only difference between CPU and GPU launcher is this kernel function, can we share other code and invoke different kernel depend on device type?

Reply:
@grimoire As far as I know, CPU/GPU functions can't be shared. Cuda kernel to specify __global__ and functions need to add __device__ specifier (CPU functions can skip this). ROI align is another example for both GPU/CPU support and is implemented in the same fashion for this reason I think.

Reply:
Errr, I means:
```c++
// modulated_deform_conv_forward.cpp

// declare cuda kernel wrap function here
void modulated_deformable_im2col_cuda();

// declare cpp implement here
void modulated_deformable_im2col_cpu();

void modulated_deform_conv_forward(){
....

// choice different function
  if(use_cuda){
    modulated_deformable_im2col_cuda();
  }else{
    modulated_deformable_im2col_cpu();
  }

....
}
```

You can keep `modulated_deformable_im2col_cuda` inside `modulated_deform_conv_cuda.cu` and invoke cuda kernel there.

Reply:
Calling 2 kernel implementations are quite different. Cuda specific syntaxes (e.g. GET_BLOCKS, at::cuda::getCurrentCUDAStream()) in the wrap function (e.g. `modulated_deformable_im2col_cuda`) should be defined in a `.cu` file. I don't think we can define both wrap functions in a cpp file. This is to ensure that each compiler (gcc, nvcc) takes care of files it knows best to compile.

Reply:
You can put the `declaration` in `cpp` and `definition` in `.cu`. Here is an example:
[link](https://drive.google.com/file/d/1JuDYHWihxrrH3JJjzpve7k1ZbtZdygUV/view?usp=sharing)

Reply:
Thanks, @grimoire. I've refactored DCN for reference. Let me know how it looks and I can refactor modulated DCN as well.

Reply:
Looks great!

Reply:
Modulated DCN is refactored accordingly as well